### PR TITLE
2707 recruitment deletemilestone endpoint

### LINKED ISSUE #2707: [Recruitment] - deleteMilestone endpoint
### Desired Changes

Blocked by #2699 

We need to be able to delete milestones. To do this we will need an endpoint. It should be under the route /recruitment/milestone/:milestoneId/delete. Only admins should be able to delete milestones.

### Acceptance Criteria
- There is a new endpoint under the route /recruitment/milestone/:milestoneId/delete
- Only admins can delete timeline events
- Unit tests are added to ensure everything works correctly
- Screenshot of manual tests from postman

### Screenshots (as needed)

_No response_

### PR DESCRIPTION
**Pull request by AShyamkumar19**
## Changes

Deletes a milestone based on the milestone id and organization id

## Screenshots
Screenshot 2024-08-07 164057

Screenshot 2024-08-07 163234

image

image

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #2707

### CODE REVIEW

File: src/backend/tests/test-utils.ts [hunk 1]
```diff
@@ -195,6 +195,22 @@ export const createTestWorkPackageTemplate = async (user: User, organizationId?:
   return workPackageTemplate;
 };
 
+export const createTestMileStone = async (user: User, organizationId: string) => {
```
Comment (Aaryan1203) [inline src/backend/tests/test-utils.ts]: createTestMilestone*

File: src/backend/tests/unmocked/recruitment.test.ts [hunk 2]
```diff
@@ -88,4 +89,41 @@ describe('Recruitment Tests', () => {
...
+    it('Fails if user is not admin', async () => {
+      await expect(
+        async () => await RecruitmentServices.deleteMilestone(await createTestUser(wonderwomanGuest, orgId), 'id', orgId)
+      ).rejects.toThrow(new AccessDeniedAdminOnlyException('delete milestone'));
+    });
+
+    it('Fails if milestoneId is not found', async () => {
+      await expect(
+        async () => await RecruitmentServices.deleteMilestone(await createTestUser(batmanAppAdmin, orgId), 'id1', orgId)
+      ).rejects.toThrow(new HttpException(400, 'Milestone with id: id1 not found!'));
```
Comment (Aaryan1203) [inline src/backend/tests/unmocked/recruitment.test.ts:258]: thijs should throw a not found exception

File: src/backend/tests/unmocked/recruitment.test.ts [hunk 3]
```diff
@@ -88,4 +89,41 @@ describe('Recruitment Tests', () => {
...
+      await RecruitmentServices.deleteMilestone(testSuperman, testMilestone.milestoneId, orgId);
+
+      await expect(
+        async () => await RecruitmentServices.deleteMilestone(testSuperman, testMilestone.milestoneId, orgId)
+      ).rejects.toThrow(new DeletedException('Milestone', testMilestone.milestoneId));
+    });
+
+    it('Succeeds and deletes milestone', async () => {
+      const testSuperman = await createTestUser(supermanAdmin, orgId);
+      const [testMilestone1] = await Promise.all([createTestMileStone(testSuperman, orgId)]);
```
Comment (Aaryan1203) [inline src/backend/tests/unmocked/recruitment.test.ts]: there is no need to make this an array and use promise.all. instead do it how you did it above on line 108

**Review by Aaryan1203 — CHANGES_REQUESTED** [review]
Great job! your tests look rlly good. Just a couple things and then it looks good for me

**Review by superhvarn — DISMISSED** [review]
lgtm

File: src/backend/src/services/recruitment.services.ts [hunk 4]
```diff
@@ -61,4 +61,28 @@ export default class RecruitmentServices {
...
   }
+
+  /**
+   * Deletes the milestone for the given milestoneId and organizationId
+   * @param deleter the name of the user deleting the milestone
+   * @param milestoneId milestone id for the specific milestone
+   * @param organizationId organization Id of the milestone
+   */
+  static async deleteMilestone(deleter: User, milestoneId: string, organizationId: string): Promise<void> {
+    if (!organizationId) throw new NotFoundException('Organization', organizationId);
```
Comment (Peyton-McKee) [inline src/backend/src/services/recruitment.services.ts]: You need to actually try to get the organization. Not just check if the orgid exists

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]

File: src/backend/src/routes/recruitment.routes.ts [hunk 5]
```diff
@@ -17,6 +17,10 @@ recruitmentRouter.post(
   RecruitmentController.createMilestone
 );
 
+recruitmentRouter.delete('/milestone/:milestoneId/delete', RecruitmentController.deleteMilestone);
+
+recruitmentRouter.get('/milestones', RecruitmentController.getAllMilestones);
```
Comment (Aaryan1203) [inline src/backend/src/routes/recruitment.routes.ts]: remove this its already there, i think it got added when we did merge confclits

File: src/backend/tests/unmocked/recruitment.test.ts [hunk 6]
```diff
@@ -187,30 +191,67 @@ describe('Recruitment Tests', () => {
     });
   });
 
-  describe('Create FAQ', () => {
-    it('Fails if user is not an admin', async () => {
+  describe('Delete a single milestone', () => {
```
Comment (Aaryan1203) [inline src/backend/tests/unmocked/recruitment.test.ts:248]: add a test for if the organiztion doesn't exist

**Review by Aaryan1203 — CHANGES_REQUESTED** [review]
couple comments but rlly nice job

**Review by Aaryan1203 — DISMISSED** [review]
lgtm

File: src/backend/src/services/recruitment.services.ts [hunk 7]
```diff
@@ -110,6 +110,35 @@ export default class RecruitmentServices {
     return allMilestones;
   }
 
+  /**
+   * Deletes the milestone for the given milestoneId and organizationId
+   * @param deleter the name of the user deleting the milestone
```
Comment (walker-sean) [inline src/backend/src/services/recruitment.services.ts]: This isn't a name, it's the user

**Review by walker-sean — CHANGES_REQUESTED** [review]
Quick comment + merge conflicts

**Pull request merged**